Honor topvalue while determining isMissingvalueCompetitive in case bottom is not set

gashutos opened the pull request:
### Description

There is a bug introduced with #12334 when field contains missing value and if those missing values are competitive.
`isMissingValueCompetitive()` implementation gets invoked comparing `missingValue` with `bottom` when bottom is not even set when `After` is set. 

We should check with topValue in cases when `bottom` is not set. This PR adds that check.
Added tests for this scenario as well. 

### Lucene issue
#12521

gashutos commented:
This was observed here -> https://github.com/opensearch-project/OpenSearch/issues/9537

gashutos committed:
Honor topvalue while determining isMissingvalueCompetitive in case bottom is not set

Signed-off-by: gashutos <gashutos@amazon.com>

jpountz reviewed (approved):
Thanks for catching this bug, the change looks correct to me.

backslasht reviewed (changes requested):
Looks good, can you add a test for missing ~~bottom~~ minimum value (for desc sort) as well?

gashutos commented:
> Looks good, can you add a test for missing bottom value as well?

Thanks @backslasht . 
Yes the missing value comparison with bottom (as missing value) is in this test.
https://github.com/apache/lucene/blob/34523ec542c21364fffefe726e6226d7f8ffa923/lucene/core/src/test/org/apache/lucene/search/TestSortOptimization.java#L219

The test missing was, missing value comparison with top value when bottom is not set and top value is coming as missing value which I added.

backslasht commented:
> > Looks good, can you add a test for missing ~~bottom~~ minimum value for desc sort as well?
> 
> Thanks @backslasht . Yes the missing value comparison with bottom (as missing value) is in this test already.
> 
> https://github.com/apache/lucene/blob/34523ec542c21364fffefe726e6226d7f8ffa923/lucene/core/src/test/org/apache/lucene/search/TestSortOptimization.java#L219
> 
> The test missing was, missing value comparison with top value when bottom is not set and top value is coming as missing value which I added.

Sorry, I jumbled the terminologies, I meant a test with missing minimum value for desc sort.

gashutos committed:
Adding missing value competitiveness check for desc order sort with after as missing value

Signed-off-by: gashutos <gashutos@amazon.com>

gashutos commented:
> > > Looks good, can you add a test for missing ~bottom~ minimum value for desc sort as well?
> > 
> > 
> > Thanks @backslasht . Yes the missing value comparison with bottom (as missing value) is in this test already.
> > https://github.com/apache/lucene/blob/34523ec542c21364fffefe726e6226d7f8ffa923/lucene/core/src/test/org/apache/lucene/search/TestSortOptimization.java#L219
> > 
> > The test missing was, missing value comparison with top value when bottom is not set and top value is coming as missing value which I added.
> 
> Sorry, I jumbled the terminologies, I meant a test with missing minimum value for desc sort.

Got it, makes sense. Added the test. I kept missing value as Long.MAX_VALUE only for this, it doesnt really matter for our logic for checking its competitiveness.

gashutos committed:
Correcting comments

Signed-off-by: gashutos <gashutos@amazon.com>

backslasht reviewed (approved):
Thanks @gashutos for making the changes.

jpountz commented:
I'll merge next week if nobody beats me to it.

jpountz commented on lucene/CHANGES.txt:
```diff
@@ -99,6 +99,8 @@ Bug Fixes
 
 * GITHUB#12220: Hunspell: disallow hidden title-case entries from compound middle/end
 
+* LUCENE-12521: Sort After returning in-correct result when missing values are competitive
```
This is under 10.0, can you move it under 9.8, IMO this change should be backported?

gashutos committed:
Update CHANGES.txt

jpountz commented:
Thanks @gashutos !
